File: c++/leetcode/3. Longest Substring Without Repeating Characters.cpp
class Solution
{
public:
  int lengthOfLongestSubstring(string s)
  {
    int max = 0;
    int right = 0;
    int left = 0;

    set<char> seen;

    while (right < s.length())
    {
      auto pos = seen.find(s[right]);
      if (pos != seen.end())
      {
        // remove val from set
        // increase left
        seen.erase(s[left]);
        left++;
      }
      else
      {
        seen.insert(s[right]);
        max = seen.size() > max ? seen.size() : max;
        right++;
      }
    }

    return max;
  }
};

File: javascript/leetcode/3. Longest Substring Without Repeating Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
  let max = 0;
  let left = 0;
  let right = 0;

  const seen = new Set();

  while (right < s.length) {
    if (seen.has(s[right])) {
      seen.delete(s[left]);
      left++;
    } else {
      seen.add(s[right]);
      max = seen.size > max ? seen.size : max;
      right++;
    }
  }

  return max;
};
